File: cryptoheir-rs/src/crypto.rs
//! Cryptographic operations for signing transactions

use crate::{
    types::{SignedTx, TransactionData, TransactionMode, TxParams},
    Result,
};
use alloy::{
    consensus::{SignableTransaction, TxEip1559, TxLegacy, TxEnvelope},
    eips::eip2718::Encodable2718,
    primitives::{Bytes, TxKind, U256},
    signers::{local::PrivateKeySigner, Signer},
};

/// Sign a transaction with a private key
pub async fn sign_transaction(tx_params: &TxParams, private_key: &str) -> Result<SignedTx> {
    // Parse private key
    let signer: PrivateKeySigner = private_key.parse()?;

    // Verify the signer address matches the from address
    let signer_address = signer.address();
    if signer_address != tx_params.transaction.from {
        return Err(eyre::eyre!(
            "Private key address {} does not match transaction from address {}",
            signer_address,
            tx_params.transaction.from
        ));
    }

    // Create and sign the transaction based on type
    let (signed_tx_bytes, tx_hash) = match tx_params.transaction.tx_type {
        2 => sign_eip1559(&tx_params.transaction, &signer).await?,
        0 => sign_legacy(&tx_params.transaction, &signer).await?,
        _ => {
            return Err(eyre::eyre!(
                "Unsupported transaction type: {}",
                tx_params.transaction.tx_type
            ))
        }
    };

    // Calculate predicted contract address for deployments
    let predicted_contract_address = if matches!(tx_params.mode, TransactionMode::Deploy) {
        Some(signer_address.create(tx_params.transaction.nonce))
    } else {
        None
    };

    // Create updated metadata
    let mut metadata = tx_params.metadata.clone();
    metadata.signed = true;
    metadata.signed_at = Some(chrono::Utc::now().to_rfc3339());

    Ok(SignedTx {
        signed_transaction: signed_tx_bytes,
        tx_hash,
        mode: tx_params.mode.clone(),
        from: tx_params.transaction.from,
        predicted_contract_address,
        metadata,
    })
}

/// Sign an EIP-1559 (Type 2) transaction
async fn sign_eip1559(
    tx_data: &TransactionData,
    signer: &PrivateKeySigner,
) -> Result<(Bytes, alloy::primitives::TxHash)> {
    let max_fee_per_gas = tx_data
        .max_fee_per_gas
        .ok_or_else(|| eyre::eyre!("max_fee_per_gas required for EIP-1559"))?;
    let max_priority_fee_per_gas = tx_data
        .max_priority_fee_per_gas
        .ok_or_else(|| eyre::eyre!("max_priority_fee_per_gas required for EIP-1559"))?;

    let tx = TxEip1559 {
        chain_id: tx_data.chain_id,
        nonce: tx_data.nonce,
        gas_limit: tx_data.gas_limit.to::<u64>(),
        max_fee_per_gas: max_fee_per_gas.to::<u128>(),
        max_priority_fee_per_gas: max_priority_fee_per_gas.to::<u128>(),
        to: tx_data.to.map(TxKind::Call).unwrap_or(TxKind::Create),
        value: tx_data.value.unwrap_or(U256::ZERO),
        input: tx_data.data.clone(),
        access_list: Default::default(),
    };

    // Get the signature hash
    let tx_hash = tx.signature_hash();

    // Sign the hash with the private key
    let sig_hash = tx.signature_hash();
    let signature = signer.sign_hash(&sig_hash).await?;

    // Create signed transaction and wrap in TxEnvelope
    let signed_tx = tx.into_signed(signature);
    let envelope = TxEnvelope::Eip1559(signed_tx);

    // Encode using alloy's encoding
    let encoded = envelope.encoded_2718();

    Ok((encoded.into(), tx_hash))
}

/// Sign a legacy (Type 0) transaction
async fn sign_legacy(
    tx_data: &TransactionData,
    signer: &PrivateKeySigner,
) -> Result<(Bytes, alloy::primitives::TxHash)> {
    let gas_price = tx_data
        .gas_price
        .ok_or_else(|| eyre::eyre!("gas_price required for legacy transaction"))?;

    let tx = TxLegacy {
        chain_id: Some(tx_data.chain_id),
        nonce: tx_data.nonce,
        gas_price: gas_price.to::<u128>(),
        gas_limit: tx_data.gas_limit.to::<u64>(),
        to: tx_data.to.map(TxKind::Call).unwrap_or(TxKind::Create),
        value: tx_data.value.unwrap_or(U256::ZERO),
        input: tx_data.data.clone(),
    };

    // Get the signature hash
    let tx_hash = tx.signature_hash();

    // Sign the hash with the private key
    let sig_hash = tx.signature_hash();
    let signature = signer.sign_hash(&sig_hash).await?;

    // Create signed transaction and wrap in TxEnvelope
    let signed_tx = tx.into_signed(signature);
    let envelope = TxEnvelope::Legacy(signed_tx);

    // Encode using alloy's encoding
    let encoded = envelope.encoded_2718();

    Ok((encoded.into(), tx_hash))
}

File: cryptoheir-go/internal/crypto/crypto.go
package crypto

import (
	"crypto/ecdsa"
	"fmt"
	"math/big"
	"strings"

	"github.com/cryptoheirdotio/cryptoheir/cryptoheir-go/internal/types"
	"github.com/ethereum/go-ethereum/common"
	coretypes "github.com/ethereum/go-ethereum/core/types"
	ethcrypto "github.com/ethereum/go-ethereum/crypto"
	"github.com/ethereum/go-ethereum/rlp"
)

// SignTransaction signs a transaction with a private key
func SignTransaction(txParams *types.TxParams, privateKeyHex string) (*types.SignedTx, error) {
	// Parse private key
	privateKeyHex = strings.TrimPrefix(privateKeyHex, "0x")
	privateKey, err := ethcrypto.HexToECDSA(privateKeyHex)
	if err != nil {
		return nil, fmt.Errorf("invalid private key: %w", err)
	}

	// Verify signer address matches from address
	publicKey := privateKey.Public()
	publicKeyECDSA, ok := publicKey.(*ecdsa.PublicKey)
	if !ok {
		return nil, fmt.Errorf("failed to cast public key to ECDSA")
	}
	signerAddress := ethcrypto.PubkeyToAddress(*publicKeyECDSA)

	if signerAddress != txParams.Transaction.From {
		return nil, fmt.Errorf("signer address %s does not match transaction from address %s",
			signerAddress.Hex(), txParams.Transaction.From.Hex())
	}

	// Sign based on transaction type
	var signedTxBytes []byte
	var txHash common.Hash

	if txParams.Transaction.TxType == 2 {
		// EIP-1559 transaction
		signedTxBytes, txHash, err = signEIP1559(&txParams.Transaction, privateKey)
		if err != nil {
			return nil, fmt.Errorf("failed to sign EIP-1559 transaction: %w", err)
		}
	} else if txParams.Transaction.TxType == 0 {
		// Legacy transaction
		signedTxBytes, txHash, err = signLegacy(&txParams.Transaction, privateKey)
		if err != nil {
			return nil, fmt.Errorf("failed to sign legacy transaction: %w", err)
		}
	} else {
		return nil, fmt.Errorf("unsupported transaction type: %d", txParams.Transaction.TxType)
	}

	// Predict contract address for deployments
	var predictedAddr *common.Address
	if txParams.Mode == types.TransactionModeDeploy {
		addr := ethcrypto.CreateAddress(txParams.Transaction.From, txParams.Transaction.Nonce)
		predictedAddr = &addr
	}

	// Build SignedTx result
	signedTx := &types.SignedTx{
		SignedTransaction:        signedTxBytes,
		TxHash:                   txHash,
		Mode:                     txParams.Mode,
		From:                     txParams.Transaction.From,
		PredictedContractAddress: predictedAddr,
		Metadata:                 txParams.Metadata,
	}

	return signedTx, nil
}

// signEIP1559 signs an EIP-1559 transaction
func signEIP1559(txData *types.TransactionData, privateKey *ecdsa.PrivateKey) ([]byte, common.Hash, error) {
	// Validate EIP-1559 fields
	if txData.MaxFeePerGas == nil || txData.MaxPriorityFeePerGas == nil {
		return nil, common.Hash{}, fmt.Errorf("EIP-1559 transaction requires max_fee_per_gas and max_priority_fee_per_gas")
	}

	// Build transaction value (default to 0)
	value := big.NewInt(0)
	if txData.Value != nil {
		value = txData.Value.ToBigInt()
	}

	// Create EIP-1559 transaction
	tx := coretypes.NewTx(&coretypes.DynamicFeeTx{
		ChainID:   big.NewInt(int64(txData.ChainID)),
		Nonce:     txData.Nonce,
		GasTipCap: txData.MaxPriorityFeePerGas.ToBigInt(),
		GasFeeCap: txData.MaxFeePerGas.ToBigInt(),
		Gas:       txData.GasLimit.ToBigInt().Uint64(),
		To:        txData.To,
		Value:     value,
		Data:      txData.Data,
	})

	// Sign the transaction
	signer := coretypes.NewLondonSigner(big.NewInt(int64(txData.ChainID)))
	signedTx, err := coretypes.SignTx(tx, signer, privateKey)
	if err != nil {
		return nil, common.Hash{}, fmt.Errorf("failed to sign transaction: %w", err)
	}

	// Encode to binary (EIP-2718 format)
	signedTxBytes, err := signedTx.MarshalBinary()
	if err != nil {
		return nil, common.Hash{}, fmt.Errorf("failed to encode signed transaction: %w", err)
	}

	return signedTxBytes, signedTx.Hash(), nil
}

// signLegacy signs a legacy (pre-EIP-1559) transaction
func signLegacy(txData *types.TransactionData, privateKey *ecdsa.PrivateKey) ([]byte, common.Hash, error) {
	// Validate legacy fields
	if txData.GasPrice == nil {
		return nil, common.Hash{}, fmt.Errorf("legacy transaction requires gas_price")
	}

	// Build transaction value (default to 0)
	value := big.NewInt(0)
	if txData.Value != nil {
		value = txData.Value.ToBigInt()
	}

	// Create legacy transaction
	tx := coretypes.NewTx(&coretypes.LegacyTx{
		Nonce:    txData.Nonce,
		GasPrice: txData.GasPrice.ToBigInt(),
		Gas:      txData.GasLimit.ToBigInt().Uint64(),
		To:       txData.To,
		Value:    value,
		Data:     txData.Data,
	})

	// Sign with EIP-155 (chain ID for replay protection)
	signer := coretypes.NewEIP155Signer(big.NewInt(int64(txData.ChainID)))
	signedTx, err := coretypes.SignTx(tx, signer, privateKey)
	if err != nil {
		return nil, common.Hash{}, fmt.Errorf("failed to sign transaction: %w", err)
	}

	// Encode to binary
	signedTxBytes, err := signedTx.MarshalBinary()
	if err != nil {
		return nil, common.Hash{}, fmt.Errorf("failed to encode signed transaction: %w", err)
	}

	return signedTxBytes, signedTx.Hash(), nil
}

// VerifySignature verifies a signed transaction matches expected parameters
func VerifySignature(signedTxBytes []byte, expectedFrom common.Address) error {
	tx := new(coretypes.Transaction)
	if err := tx.UnmarshalBinary(signedTxBytes); err != nil {
		return fmt.Errorf("failed to decode transaction: %w", err)
	}

	// Extract signer address from signature
	chainID := tx.ChainId()
	var signer coretypes.Signer
	if tx.Type() == coretypes.DynamicFeeTxType {
		signer = coretypes.NewLondonSigner(chainID)
	} else {
		signer = coretypes.NewEIP155Signer(chainID)
	}

	from, err := coretypes.Sender(signer, tx)
	if err != nil {
		return fmt.Errorf("failed to recover signer: %w", err)
	}

	if from != expectedFrom {
		return fmt.Errorf("signature verification failed: expected %s, got %s", expectedFrom.Hex(), from.Hex())
	}

	return nil
}

// PredictContractAddress predicts the address of a contract deployment
func PredictContractAddress(deployer common.Address, nonce uint64) common.Address {
	return ethcrypto.CreateAddress(deployer, nonce)
}

// RLPEncode encodes data using RLP encoding
func RLPEncode(data interface{}) ([]byte, error) {
	return rlp.EncodeToBytes(data)
}
